File: src/main/java/io/github/syst3ms/skriptparser/classes/PatternType.java
package io.github.syst3ms.skriptparser.classes;

/**
 * A type used in a pattern.
 * Groups a {@link Type} and a number together (in contrast to {@link Type} itself)
 */
public class PatternType<T> {
	private Type<T> type;
	private boolean single;

	public PatternType(Type<T> type, boolean single) {
		this.type = type;
		this.single = single;
	}

	public Type<T> getType() {
		return type;
	}

	public boolean isSingle() {
		return single;
	}

	@Override
	public boolean equals(Object obj) {
		if (obj == null || !(obj instanceof PatternType)) {
			return false;
		} else {
			PatternType<?> o = (PatternType<?>) obj;
			return type.equals(o.type) && single == o.single;
		}
	}

    @Override
    public String toString() { // Not perfect, but good enough for toString()
        String baseName = type.getBaseName();
        if (baseName.endsWith("child")) { // This exception seems likely enough
            return baseName.replace("child", "children");
        } else if (baseName.matches(".*(s|ch|sh|x|z)")) {
            return baseName + "es";
        } else {
            return baseName + "s";
        }
    }
}

File: skript-parser-kotlin/src/main/kotlin/io/github/syst3ms/skriptparser/classes/PatternType.kt
package io.github.syst3ms.skriptparser.classes

import io.github.syst3ms.skriptparser.registration.Type

/**
 * A type used in a pattern.
 * Groups a [Type] and a number together (in contrast to [Type] itself)
 */
class PatternType<out T>(private val type: Type<out T>, private val isSingle: Boolean) {

    override fun equals(other: Any?): Boolean {
        return if (other == null || other !is PatternType<*>) {
            false
        } else {
            type == other.type && isSingle == other.isSingle
        }
    }

    override fun toString(): String { // Not perfect, but good enough for toString()
        val baseName = type.baseName
        return when {
            baseName.endsWith("child") -> // This exception seems likely enough
                baseName.replace("child", "children")
            ".*(s|ch|sh|x|z)".toRegex().matches(baseName) -> baseName + "es"
            else -> baseName + "s"
        }
    }

    override fun hashCode(): Int {
        var result = type.hashCode()
        result = 31 * result + isSingle.hashCode()
        return result
    }
}
